### ISSUE TITLE
Introduce IReadOnlyObservableCollection<T>, a common interface for observable collections

### ISSUE THREAD

**Issue by jakub-klima** [dotnet/runtime#36945]
`IReadOnlyObservableCollection<T>` would be a common read-only interface for `ObservableCollection<T>` and `ReadOnlyObservableCollection<T>` (and similar custom implementations). It would compose already implemented interfaces into a single one. That allows to accept all its implementations into a method or to express read-only intention when working with `ObservableCollection<T>`.

The proposed `IReadOnlyObservableCollection<T>` implemented by `ObservableCollection<T>` and `ReadOnlyObservableCollection<T>` seems consistent with existing `IReadOnlyList<T>` implemented by `List<T>` and `ReadOnlyCollection<T>`.

## Proposed API

``` C#
public interface IReadOnlyObservableCollection<T> : IReadOnlyList<T>, INotifyCollectionChanged, INotifyPropertyChanged
```

```diff
-public class ObservableCollection<T> : Collection<T>, INotifyCollectionChanged, INotifyPropertyChanged
+public class ObservableCollection<T> : Collection<T>, IReadOnlyObservableCollection<T>
```

```diff
-public class ReadOnlyObservableCollection<T> : ReadOnlyCollection<T>, INotifyCollectionChanged, INotifyPropertyChanged
+public class ReadOnlyObservableCollection<T> : ReadOnlyCollection<T>, IReadOnlyObservableCollection<T>
```

It seems others would also appreciate this change:
* https://developercommunity.visualstudio.com/content/problem/327755/iobservablecollection-ireadonlyobservablecollectio.html
* https://github.com/github/VisualStudio/blob/master/src/GitHub.Extensions/IReadOnlyObservableCollection.cs
* https://stackoverflow.com/questions/1763696/how-can-i-make-a-read-only-observablecollection-property#comment50285595_15474416

**Comment by omariom** [comment]
ReadOnlyObservableCollection<T> and ObservableCollection<T> are most often not passed into methods, but exposed from object models as properties.  So IReadOnlyObservableCollection<T> is less useful in the scenario as we don't want the consumers of the model to just cast it to ObservableCollection<T> and mute it.

**Comment by jakub-klima** [comment]
I agree that if you are not sure what clients do (e.g. in public API) you should use read-only wrapper (use case 1). But for internal purposes where you know your clients I think using the interface might be easier than wrapping it (use case 2). (Also, can't there be a performance problem with wrapping in some scenario?) So, are there any downsides to add the interface to help with use case 2?

**Comment by jakub-klima** [comment]
Also maybe there is another case where interface is useful. What about unified working with custom observable collections? If I'm not mistaken, `ReadOnlyObservableCollection<T>` is made just for `ObservableCollection<T>`, and so custom implementation could not be easily wrapped in it.

**Comment by jakub-klima** [comment]
Implementation here if API got approved: https://github.com/jakub-klima/runtime/commit/c411353a3c65c265d9ada592b2f963b0d3d7b9a7

**Comment by eiriktsarpalis** [comment]
I would be interested in understanding the motivation for adding these (perhaps by including a few code samples, and how they could be currently worked around). 

Interestingly, I was going through the `github/VisualStudio` repo that you linked to and discovered [this comment](https://github.com/github/VisualStudio/blob/3b49cdbd8bfdb03f2177d32df0d72cf551525198/src/GitHub.Extensions/ObservableCollectionEx.cs#L11-L13). If the motivation for adding the interface is the assertion that `ReadOnlyObservableCollection<T>` does not implement `INotifyCollectionChanged`, then this is either incorrect or was addressed recently.

**Comment by jakub-klima** [comment]
The motivation is different than mentioned in `github/VisualStudio`. I will try to illustrate on code sample:

```
 1. DoStuff(observableCollection)
 2. {
 3.     // process items via IEnumerable
 4.     // attach on event via INotifyCollectionChanged
 5.     // e.g. https://github.com/jakub-klima/px-work-log/blob/master/PxWorkLog.Core/Utils/CollectionEventAggregator.cs
 5. }
 6.
 7. ...
 8.
 9. class MyReadOnlyObservableCollection : IEnumerable, INotifyCollectionChanged
10. {
11.     // e.g. https://github.com/jakub-klima/px-work-log/blob/master/PxWorkLog.Core/Utils/ProjectedObservableCollection.cs
12. }
13. 
14. ...
15.
16. var collection1 = new ObservableCollection();
17. var collection2 = new ReadOnlyObservableCollection();
18. var collection3 = new MyReadOnlyObservableCollection();
19.
20. DoStuff(collection1);
21. DoStuff(collection2);
22. DoStuff(collection3);
```

Solution if we have IReadOnlyObservableCollection:

- line 1: let DoStuff() accept IReadOnlyObservableCollection
- line 9: let MyReadOnlyObservableCollection implement IReadOnlyObservableCollection

Workaround 1:

- line 1: let DoStuff() accept ReadOnlyObservableCollection
- line 20: use collection1.AsReadOnly() -- extra method call overhead
- line 22: use collection3.AsReadOnly() -- AFAI **cannot be done** as ReadOnlyObservableCollection has only constructor accepting ObservableCollection

Workaround 2:

- line 1: let DoStuff() accept two arguments to pass one collection as both IEnumerable and INotifyCollectionChanged
- line 20: DoStuff(collection1, collection1)
- line 21: DoStuff(collection2, collection2)
- line 22: DoStuff(collection3, collection3) -- not very elegant

**Comment by eiriktsarpalis** [comment]
Right, so it's about dependency inversion.

A potential third work around is to create a custom `IReadOnlyObservableCollection` interface in your code and create a wrapper class for `ReadOnlyObservableCollection`. You could then write:
```csharp
var collection1 = new ObservableCollection();
var collection2 = new ReadOnlyObservableCollection();
var collection3 = new MyReadOnlyObservableCollection();

DoStuff(collection1.ToMyCustomInterface()); // extension methods that map to the wrapper
DoStuff(collection2.ToMyCustomInterface()); // »
DoStuff(collection3); // implements the custom interface directly
```

I'm slightly concerned about adding this interface since there would only be one type implementing it: `ReadOnlyObservableCollection` (and arguably also `ObservableCollection` which is essentially the same implementation). We'd need strong evidence that whatever abstraction we end up inserting is really general-purpose enough to satisfy most observable collection requirements.

That being said, the proposal itself is pretty self-consistent, and we have evidence that a partner team used the exact same design when building an interface of their own. There is always a runtime cost associated with adding more interfaces to a class, however the class itself is not used commonly enough to warrant significant concern.

**Comment by jbe2277** [comment]
I have written some lines why I believe exposing a collection via interface is a good approach:

**How should a class expose a collection?**
https://github.com/jbe2277/waf/wiki/How-should-a-class-expose-a-collection%3F

Additionally, My library `System.Waf.Core` provides the proposed interface: `IReadOnlyObservableList`
https://github.com/jbe2277/waf/blob/master/src/System.Waf/System.Waf/System.Waf.Core/Foundation/IReadOnlyObservableList.cs
